### PR TITLE
C++ API parity: BCEWithLogitsLoss

### PR DESCRIPTION
**Pull request by pbelevich**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28783 C++ API parity: BCEWithLogitsLoss**

Differential Revision: [D18202435](https://our.internmc.facebook.com/intern/diff/D18202435)

### CODE REVIEW

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 1]
```diff
@@ -232,5 +232,24 @@ Tensor PoissonNLLLossImpl::forward(
   return F::poisson_nll_loss(log_input, target, options);
 }
 
+// ============================================================================
+
+BCEWithLogitsLossImpl::BCEWithLogitsLossImpl(
+  const BCEWithLogitsLossOptions& options_) : options(options_) {
+    register_buffer("weight",options.weight());
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: nit:
```suggestion
    register_buffer("weight", options.weight());
```

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 2]
```diff
@@ -155,6 +155,20 @@ inline Tensor poisson_nll_loss(const Tensor& input, const Tensor& target,
...
+  const BCEWithLogitsLossOptions& options = {}) {
+
+  TORCH_CHECK(target.sizes() == input.sizes(),
+    "Target size (", target.sizes(),
+    ") must be the same as input size (",
+    input.sizes(), ")"
+  );
+
+  return torch::binary_cross_entropy_with_logits(input, target,
+    options.weight(), options.pos_weight(), get_enum(options.reduction()));
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: I believe we don't need to call `get_enum` because the value of `options.reduction()` already maps to the correct enum value that `torch::binary_cross_entropy_with_logits` accepts.

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 3]
```diff
@@ -334,5 +334,27 @@ Tensor CrossEntropyLossImpl::forward(
...
+// ============================================================================
+
+BCEWithLogitsLossImpl::BCEWithLogitsLossImpl(
+  const BCEWithLogitsLossOptions& options_) : options(options_) {
+  reset();
+}
+
+void BCEWithLogitsLossImpl::reset() {
+  register_buffer("weight", options.weight());
+  register_buffer("pos_weight", options.pos_weight());
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: In Python, I think registering a buffer makes it available through class member accessor such as `module.weight` and `module.pos_weight`, and to mimic this in C++ version I think we might want to have `weight` and `pos_weight` in the `BCEWithLogitsLossImpl` class, and here we can do
```cpp
weight = register_buffer("weight", options.weight());
pos_weight = register_buffer("pos_weight", options.pos_weight());
```

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 4]
```diff
@@ -589,6 +593,32 @@ inline Tensor cross_entropy(
       options.reduction());
 }
 
+// ============================================================================
+
+namespace detail {
+inline Tensor binary_cross_entropy_with_logits(
+  const Tensor& input, const Tensor& target,
+  const Tensor& weight, const Tensor& pos_weight,
+  BCEWithLogitsLossOptions::reduction_t reduction) {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: To follow the ordering in the Python version, maybe `reduction` before `pos_weight`?

File: torch/csrc/api/include/torch/nn/options/loss.h [hunk 5]
```diff
@@ -304,5 +304,22 @@ struct TORCH_API CrossEntropyLossOptions {
...
+  /// If given, has to be a Tensor of size `nbatch`.
+  TORCH_ARG(Tensor, weight) = {};
+  /// Specifies the reduction to apply to the output. Default: Mean
+  TORCH_ARG(reduction_t, reduction) = torch::kMean;
+  /// A weight of positive examples.
+  /// Must be a vector with length equal to the number of classes.
+  TORCH_ARG(Tensor, pos_weight) = {};
+};
+
+TORCH_NN_FUNCTIONAL_USE_MODULE_OPTIONS(BCEWithLogitsLoss, BCEWithLogitsLossFuncOptions)
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/loss.h]: I feel that we might want to do
```suggestion
TORCH_NN_FUNCTIONAL_USE_MODULE_OPTIONS(BCEWithLogitsLoss, BinaryCrossEntropyWithLogitsFuncOptions)
```
to more closely match the functional name `binary_cross_entropy_with_logits`.

**Review by yf225 — APPROVED** [review]
@pbelevich Thanks so much for the awesome work!

**Comment by jjlilley** [comment]
I'm seeing the following in CircleCI, re: caffe2/test/cpp/api/functional.cpp:2491

Is this related?

Nov 20 18:01:35 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "In file included from \u{1b}[01m\u{1b}[K/var/lib/jenkins/workspace/third_party/googletest/googletest/include/gtest/gtest.h:59:0\u{1b}[m\u{1b}[K,\n                 from \u{1b}[01m\u{1b}[K/var/lib/jenkins/workspace/test/cpp/api/functional.cpp:1\u{1b}[m\u{1b}[K:\n\u{1b}[01m\u{1b}[K/var/lib/jenkins/workspace/test/cpp/api/functional.cpp:\u{1b}[m\u{1b}[K In member function \'\u{1b}[01m\u{1b}[Kvirtual void FunctionalTest_BCEWithLogitsLoss_Test::TestBody()\u{1b}[m\u{1b}[K\':\n\u{1b}[01m\u{1b}[K/var/lib/jenkins/workspace/test/cpp/api/functional.cpp:2491:9:\u{1b}[m\u{1b}[K \u{1b}[01;31m\u{1b}[Kerror: \u{1b}[m\u{1b}[K\'\u{1b}[01m\u{1b}[KBCELossFuncOptions\u{1b}[m\u{1b}[K\' is not a member of \'\u{1b}[01m\u{1b}[KF\u{1b}[m\u{1b}[K\'\n         F::BCELossFuncOptions().weight(weight)\n\u{1b}[01;32m\u{1b}[K         ^\u{1b}[m\u{1b}[K\n\u{1b}[01m\u{1b}[K/var/lib/jenkins/workspace/test/cpp/api/functional.cpp:2508:9:\u{1b}[m\u{1b}[K \u{1b}[01;31m\u{1b}[Kerror: \u{1b}[m\u{1b}[K\'\u{1b}[01m\u{1b}[KBCELossFuncOptions\u{1b}[m\u{1b}[K\' is not a member of \'\u{1b}[01m\u{1b}[KF\u{1b}[m\u{1b}[K\'\n         F::BCELossFuncOptions().reduction(torch::kNone)\n\u{1b}[01;32m\u{1b}[K         ^\u{1b}[m\u{1b}[K\n\u{1b}[01m\u{1b}[K/var/lib/jenkins/workspace/test/cpp/api/functional.cpp:2518:9:\u{1b}[m\u{1b}[K \u{1b}[01;31m\u{1b}[Kerror: \u{1b}[m\u{1b}[K\'\u{1b}[01m\u{1b}[KBCELossFuncOptions\u{1b}[m\u{1b}[K\' is not a member of \'\u{1b}[01m\u{1b}[KF\u{1b}[m\u{1b}[K\'\n         F::BCELossFuncOptions().weight(weight)\n\u{1b}[01;32m\u{1b}[K         ^\u{1b}[m\u{1b}[K\n" }

**Comment by yf225** [comment]
@jjlilley Is it related to https://github.com/pytorch/pytorch/pull/30146 and I just reverted that PR. It should be green on master now :D
